File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.swift
import Foundation

func solution(_ n:Int, _ wires:[[Int]]) -> Int {
    var graph: [[Int]] = Array(repeating: [], count:n+1)
    var answer = Int(1e9)
    var visited:[Bool] = []
    
    for wire in wires {
        let a = wire[0]
        let b = wire[1]
        graph[a].append(b)
        graph[b].append(a)
    }
        
    func bfs(_ start: Int, _ link: [[Bool]]) -> Int {
        visited[start] = true
        var q = [start]
        var count = 1
        
        while !q.isEmpty {
            var x = q.removeFirst()
            
            for i in graph[x] {
                if !visited[i] && link[x][i] {
                    visited[i] = true
                    count += 1
                    q.append(i)
                }
            }
        }
        
        return count
    }
    
    var link = Array(repeating: Array(repeating: true, count:n+1), count:n+1)
    
    for wire in wires {
        let a = wire[0]
        let b = wire[1]
        
        visited = Array(repeating: false, count: n+1)
        link[a][b] = false
        let result1 = bfs(a, link)
        let result2 = bfs(b, link)
        link[a][b] = true
        
        answer = min(answer, abs(result1 - result2))
    }
    
    return answer
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    answer = 101
    graph = [[] for _ in range(n+1)]
    
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)
    
    def dfs(v):
        visited[v] = True
        cnt = 1
        
        for node in graph[v]:
            if not visited[node]:
                cnt += dfs(node)
                
        return cnt
                
    for a, b in wires:
        graph[a].remove(b)
        graph[b].remove(a)
        cnt_a = 0
        visited = [False] * (n+1)
        
        for i in range(1, n+1):
            if graph[i] and not visited[i]:
                cnt_a = dfs(i)
                
        cnt_b = n - cnt_a
        answer = min(answer, abs(cnt_b - cnt_a))
        
        graph[a].append(b)
        graph[b].append(a)
        
    return answer